Add VisualShape2D node for drawing common shapes

kitbdev opened the pull request:
- closes https://github.com/godotengine/godot-proposals/issues/1126

![Screenshot 2024-11-13 193704](https://github.com/user-attachments/assets/ebbfbf98-32d2-406d-8683-540c358fcf99)

Draw common shapes like Rectangles, Circles, Equilateral Triangles, Right Triangles, Capsules, and Regular Polygons.
Vector-based like Polygon2D, does not rely on textures.
Outline with variable thickness.
Antialiasing.
Solid color (the above image uses 2D lighting on some shapes).
UVs (except on outlines and antialiasing) for shader support.
Regular polygons are under the Circle shape by changing the resolution.

In the editor:
Convert to MeshInstance2D or Polygon2D. 
Add CollisionShape2D or ShadowMask2D sibling. 
These are similar to the functionality of Sprite.

![image](https://github.com/user-attachments/assets/aa9d3a30-5d90-4212-828a-3c1c39d84e99)

<details>
<summary>Implementation notes:</summary>

- The name Shape2D is taken so I went for VisualShape2D instead.
- I couldn't use Shape2D resources since it has Shapes that don't really make sense for drawing like separation ray or world boundary. It also didn't have all the shapes I want, I didn't want to add new ones since it affects physics workflow, and I didn't want to just hide them under convex shape somehow. Also, I wanted to not use a resource, to be more beginner friendly when duplicating nodes.
- I didn't add a Texture option to try to keep things simple, if a texture is wanted, a Sprite and clip children can be used, or a custom material.
- The canvas editor rect tool adjusts the size instead of the scale. Using a size variable is important for easier nesting of shapes with different sizes, for the capsule shape, and for antialiasing.
- ~Capsules are vertical only, with some more math it could probably dynamically be vertical or horizontal based on the size.~
- This can add CollisionShape2Ds instead of CollisionPolygon2D nodes compared to Sprite2D.
- Ovals (Circle shape with non-uniform size) will generate a circle collision shape rather than a convex one, since I didn't want it to decide which one it should do automatically. Right now they use the max, so the circle covers the entire oval.
- Outlines just use the draw method's implementation, so it has limitations. It doesn't have proper UVs and it can look off at large sizes on some shapes. When converting or adding sibling, the outline will be ignored. I also ignore the outline for the editor tools since calculating the size isn't as straightforward as just adding the outline width, especially when it gets larger on non-uniform sized shapes. Also thin outlines may be hard to click on.
- Enabling outlines will stop rendering the interior of the shape, since rather than having an outline color property, a second shape could be used to fill it.
- Antialiasing works by adding a 1px outline, since the `draw_polyline` function has it and `draw_colored_polygon` doesn't. This doesn't have UVs, and half of the line is drawn behind the shape. Alias is also affect by scaling.
- The resolution property allows the circle shape to become any regular polygon. Maybe circles and regular polygons should be separated with different default resolutions? I kept equilateral triangle separate even though it can be made with a circle as it is common and the circle shape triangle is offset from the base, which makes it a bit harder to work with.
- The default resolution is 64 since that is what `draw_circle` uses.
- Polygon2D UV data is weird, it isn't in the range 0-1 like other UV data and it seems to only apply if a texture is set, even if a material would make use of it. When generating the UVs for it I use the size of the shape, so a 128x128 texture fits a 128x128 shape.
- I added some keywords, but there might be other appropriate ones I missed.
</details>

kitbdev committed:
Add VisualShape2D node for drawing common shapes

Mickeon commented on doc/classes/VisualShape2D.xml:
```diff
@@ -0,0 +1,66 @@
+			A square or rectangle shape.
+		</constant>
+		<constant name="SHAPE_CIRCLE" value="1" enum="ShapeType">
+			A circle or regular polygon shape. When [member resolution] is high, this approximates a circle. This can be an oval if the [member size] is non-uniform.
```
Not too strong on this, feel free to dismiss.
```suggestion
			A circle or regular polygon shape. When [member resolution] is high, this approximates a circle. This can also be an oval if the [member size] is non-uniform.
```

Mickeon commented on doc/classes/VisualShape2D.xml:
```diff
@@ -0,0 +1,66 @@
+			If greater than [code]0[/code], the shape uses an outline instead of being filled. This is the outline width of the shape in pixels. UVs are not supported on outlines.
+		</member>
+		<member name="resolution" type="int" setter="set_resolution" getter="get_resolution" default="64">
+			The resolution to use when the [member shape_type] is [constant SHAPE_CIRCLE] or [constant SHAPE_CAPSULE]. The resolution is used as the number of points in the shape. Large values may affect performance.
```
```suggestion
			The resolution to use when the [member shape_type] is [constant SHAPE_CIRCLE] or [constant SHAPE_CAPSULE]. This value is the number of points in the shape. Higher values look smoother but may negatively affect performance.
```
or
```suggestion
			The number of points in the shape if the shape is [constant SHAPE_CIRCLE] or [constant SHAPE_CAPSULE]. Higher values look smoother but may negatively affect performance.
```
Or similar.
Bit of a note: a similar property in StyleBoxFlat is called `corner_detail` and refers to the points as "vertices".

kitbdev replied:
I think I'll go with the first but say "This value determines the number of points in the shape", since in capsules its not actually the number of points, its rounded up to the next even number, plus two. (since it needs to be even and the flat sides add some). I'll mention it in the capsule description too.

I could use vertices instead of point here, though I have a method `get_points`. Polygon2D seems to use them interchangeably.

Loufe replied:
"The resolution applied when the [member shape_type] is [constant SHAPE_CIRCLE] or [constant SHAPE_CAPSULE]. This value determines the number of points in the shape. Higher values appear smoother but may negatively affect performance."
My 2 cents, I'd swap out "to use" with applied and "appear" instead of "look".

Mickeon commented on doc/classes/VisualShape2D.xml:
```diff
@@ -0,0 +1,66 @@
+			An equilateral triangle shape, pointing up.
+		</constant>
+		<constant name="SHAPE_RIGHT_TRIANGLE" value="3" enum="ShapeType">
+			A right triangle shape, occupying the bottom left corner.
```
I kind of understand what this means but it could be elaborated on.

kitbdev replied:
I'm not sure how to describe it, maybe "The left and bottom sides make up the legs of the triangle." ?

![image](https://github.com/user-attachments/assets/d375ecf6-5858-4e91-9d48-1caf7ac43ade)

Mickeon replied:
```suggestion
			A right triangle shape, with its right angle placed on the node's bottom-left corner.
```
or
```suggestion
			A right triangle shape. Its right angle starts from the node's bottom-left corner.
```
I am not sure if this is a correct way of expressing this in geometry terms, but at the same time, it's very easy to imagine a right triangle.

Loufe replied:
I've always known this as a "right-angled triangle", though @kitbdev 's "right triangle" seems technically correct, as well:
https://en.wikipedia.org/wiki/Right_triangle

Mickeon commented on doc/classes/VisualShape2D.xml:
```diff
@@ -0,0 +1,66 @@
+			A right triangle shape, occupying the bottom left corner.
+		</constant>
+		<constant name="SHAPE_CAPSULE" value="4" enum="ShapeType">
+			A vertical capsule shape. The [member size] height should be greater than the width.
```
What happens when the height is not greater than the width, then?

Mickeon replied:
```suggestion
			A vertical capsule shape. The [member size] height should be greater than the width, otherwise this shape will appear as a circle.
```

Mickeon commented on doc/classes/VisualShape2D.xml:
```diff
@@ -0,0 +1,66 @@
+	</methods>
+	<members>
+		<member name="antialiased" type="bool" setter="set_antialiased" getter="is_antialiased" default="false">
+			If [code]true[/code], the shape will be antialiased. UVs are not supported on antialiased edges. Scaling may affect antialiasing quality.
```
```suggestion
			If [code]true[/code], the shape is antialiased. UVs are not supported on antialiased edges. Scaling may affect antialiasing quality.
```

Mickeon reviewed (commented):
Oh wow, I was not expecting this to be requested **this much**.

When adding new nodes like this, my only personal concern is compiled binary size.

Loufe commented on doc/classes/VisualShape2D.xml:
```diff
@@ -0,0 +1,66 @@
+		2D shape-drawing node.
+	</brief_description>
+	<description>
+		VisualShape2D is a node for drawing common geometric shapes like rectangles, circles, triangles, and capsules. For more advanced capabilities, see [Polygon2D], [Sprite2D], or [MeshInstance2D]. For custom drawing, see [method CanvasItem._draw] and the [CanvasItem] draw methods.
```
"VisualShape2D is a node supporting common geometric shapes such circles, triangles, and capsules. For use cases requiring more detail may be better served by [Polygon2D], [Sprite2D], or [MeshInstance2D]. For custom drawing, see [method CanvasItem._draw] and the [CanvasItem] draw methods."
I'm not sure of the value add by mentioning prototyping. Added some small gramatical changes to my taste.

Mickeon replied:
Rapid prototyping is one of the main reason this kind of feature is desired. It should not go unnoticed and should instead be elaborated upon.

"For use cases requiring more detail may be better served by" is unusually posh for the class reference.

Mickeon commented:
Despite the current milestone and popular demand, this feature would require a good chunk of more recent discussion. Merging for `4.4` is not as straightforward as one may think. As is the general rule of thumb, I'm going to shift this PR's milestone to `4.x`.

Bear in mind this does not mean the PR is abandoned, or undesired.

KoBeWi commented:
For reference:
Previous implementation: #16483 (rejected)
Alternative: #84587

I tested this and I like the implementation a lot. It's complete with its own editor and the shapes even work properly with shaders. The rectangle shape is a Node2D equivalent of ColorRect, so I imagine it can be used outside prototyping (previous workaround was a scaled Sprite2D with 1px texture).

There was lots of discussion in the proposal and the alternative of simply using CollisionPolygon2D or custom-drawn Shape2D is much inferior to this PR's implementation. Also this can't be easily replicated with a plugin (it's a lot of code); it could work nice as a GDExtension though.

I said I like the implementation, but that's not really enough to merge a feature. As a heavy 2D user, I could *maybe* see myself using it in some cases, but the existing workarounds are good enough. However I'm not against merging this either.

> I didn't add a Texture option to try to keep things simple, if a texture is wanted, a Sprite and clip children can be used, or a custom material.

It's pretty much given that a texture will be among the first things requested and someone will eventually add it. Every time engine gets a new thing, it's followed by various follow-up feature requests (something I observed recently).

---

Playing with resolution gives interesting results 🤔
![image](https://github.com/user-attachments/assets/3fef0b88-a60c-4bb6-8b6c-abf181a32d83)

I think the capsule should change orientation when width is bigger than height.

kitbdev commented:
Yeah, I figured I'd give it another try since the proposal seems wanted and has the implementer wanted tag.

For the Texture support, maybe we can add a note in the documentation to head that off? With texture support comes wanting UV offsets, flipping, atlas and animation support, and at that point it would be better to just inherit from Sprite2D.

Added horizontal capsules. I used an approach that makes sure the shape will be the same as the vertical one, if rotated.
However if the resolution is small and not a multiple of 4, the transition from horizontal to vertical may be noticeable. The alternative is to have the shapes not be identical when rotated, I can change to that approach if wanted. 
Also some resolutions may look a bit different than in the previous commit, due to changing the starting point, they no longer always have a point at the ends, and the width should be more consistent.
Making a CollisionShape2D based on a horizontal capsule will automatically rotate it so it matches.

![image](https://github.com/user-attachments/assets/a2e85af2-e3fb-4f3e-b7fd-cb91c13f77c9)

Also fixed issue when changing the rect in 2D editor when offset was not at the center. Offset now moves to be in the same relative position.

KoBeWi commented:
> Yeah, I figured I'd give it another try since the proposal seems wanted and has the implementer wanted tag.

Indeed, seems like it was basically accepted: https://github.com/godotengine/godot-proposals/issues/1126#issuecomment-1184568550
I missed this comment.

KoBeWi commented on doc/classes/VisualShape2D.xml:
```diff
@@ -0,0 +1,67 @@
+	</brief_description>
+	<description>
+		[VisualShape2D] is a node that draws common geometric shapes like rectangles, circles, triangles, and capsules. It is very useful to prototype 2D applications, as it needs very little setup without needing a [Texture2D].
+		For more advanced capabilities, see [Polygon2D], [Sprite2D], or [MeshInstance2D]. For custom drawing, see [method CanvasItem._draw] and the [CanvasItem] draw methods.
```
I'd also mention that it converts to various nodes.

kitbdev replied:
I worked it into a sentence about texture. 
I couldn't find too many other references to this kind of functionality, so I just said the Editor can convert it, I don't know if there is a specific term for those buttons that we want to use.

KoBeWi commented on scene/2d/visual_shape_2d.h:
```diff
@@ -0,0 +1,108 @@
+	ShapeType shape_type = SHAPE_RECTANGLE;
+	Color color = Color(1, 1, 1, 1);
+	Size2 size = Size2(128, 128);
+	Point2 offset = Point2();
```
```suggestion
	Point2 offset;
```

KoBeWi commented on scene/2d/visual_shape_2d.cpp:
```diff
@@ -0,0 +1,351 @@
+}
+
+void VisualShape2D::_edit_set_rect(const Rect2 &p_edit_rect) {
+	ERR_FAIL_COND(!_edit_use_rect());
```
This will never fail.

kitbdev replied:
I thought so, but that's what `Node2D::_edit_set_rect` does.

KoBeWi replied:
The method is virtual, so the base method can fail if inheriting class doesn't use rect.

KoBeWi commented on scene/2d/visual_shape_2d.cpp:
```diff
@@ -0,0 +1,351 @@
+#endif // DEBUG_ENABLED
+
+void VisualShape2D::_notification(int p_what) {
+	bool filled = outline_width <= 0;
```
This should be inside NOTIFICATION_DRAW.

KoBeWi commented on scene/2d/visual_shape_2d.cpp:
```diff
@@ -0,0 +1,351 @@
+}
+
+void VisualShape2D::set_color(const Color &p_color) {
+	color = p_color;
```
Missing check for the same color.

kitbdev replied:
Should I do the same for the other properties as well? `set_size`, `set_outline_width`, `set_antialiased`, etc.

KoBeWi replied:
Yes.
Any setter that has side-effect (e.g. calls `queue_redraw()`) should be guarded.

KoBeWi commented on editor/plugins/visual_shape_2d_editor_plugin.h:
```diff
@@ -0,0 +1,91 @@
+#define VISUAL_SHAPE_2D_EDITOR_PLUGIN_H
+
+#include "editor/plugins/editor_plugin.h"
+#include "scene/2d/visual_shape_2d.h"
```
VisualShape2D can be forward-declared.

KoBeWi commented on editor/plugins/visual_shape_2d_editor_plugin.cpp:
```diff
@@ -0,0 +1,292 @@
+#include "scene/resources/2d/convex_polygon_shape_2d.h"
+#include "scene/resources/2d/rectangle_shape_2d.h"
+
+void VisualShape2DEditor::_node_removed(Node *p_node) {
```
You can handle `edit(nullptr)` instead of adding this method. Should be the same.

kitbdev replied:
I copied this from Sprite2D, but I don't think this actually ever gets called, same with the one in Sprite2D.

KoBeWi replied:
Sprite2D editor is very old.

KoBeWi commented on editor/plugins/visual_shape_2d_editor_plugin.cpp:
```diff
@@ -0,0 +1,292 @@
+	selected_menu_item = (Menu)p_option;
+
+	if (visual_shape_2d->get_owner() != get_tree()->get_edited_scene_root()) {
+		EditorToaster::get_singleton()->popup_str(TTR("Can't convert a VisualShape from a foreign scene."), EditorToaster::SEVERITY_ERROR);
```
You forgot `return`.
Also creating shapes/occluders is not a conversion, so could be allowed.

kitbdev replied:
How can I check if the node is a scene root of a foreign scene?

KoBeWi replied:
I meant root of the current scene. If `visual_shape_2d` is the root, it will have null owner, so this check fails.

KoBeWi commented on editor/plugins/visual_shape_2d_editor_plugin.cpp:
```diff
@@ -0,0 +1,292 @@
+			collision_shape_2d_instance->set_shape(shape);
+			collision_shape_2d_instance->translate(visual_shape_2d->get_offset());
+		} break;
+		case VisualShape2D::SHAPE_CIRCLE: {
```
This might need a special case for ovals. It's the only shape that doesn't get accurate collider.

kitbdev replied:
I could calculate the eccentricity and make a ConvexPolygonShape2D in some cases, but I didn't want to arbitrarily choose some threshold. I also want to have it use CircleShape2D where possible, since it is better for physics.
I can play around with thresholds and see.

kitbdev replied:
Eccentricity wasn't very helpful, for now I went with a 10% difference in size.
If it still feels off, I could instead do what MeshInstance3DEditor does and have a menu and let the user choose whether to make it a shape or a convex polygon. This would also allow for small resolution shapes (like hexagon) to have a polygon option.

KoBeWi commented on editor/plugins/visual_shape_2d_editor_plugin.cpp:
```diff
@@ -0,0 +1,292 @@
+
+void VisualShape2DEditor::_notification(int p_what) {
+	switch (p_what) {
+		case NOTIFICATION_ENTER_TREE:
```
```suggestion
```

KoBeWi commented on editor/plugins/visual_shape_2d_editor_plugin.cpp:
```diff
@@ -0,0 +1,292 @@
+	ur->create_action(TTR("Create LightOccluder2D Sibling"), UndoRedo::MERGE_DISABLE, visual_shape_2d);
+	ur->add_do_method(this, "_add_as_sibling_or_child", visual_shape_2d, light_occluder_2d_instance);
+	ur->add_do_reference(light_occluder_2d_instance);
+	ur->add_undo_method(visual_shape_2d != get_tree()->get_edited_scene_root() ? visual_shape_2d->get_parent() : get_tree()->get_edited_scene_root(), "remove_child", light_occluder_2d_instance);
```
```suggestion
	ur->add_undo_method(visual_shape_2d != get_tree()->get_edited_scene_root() ? visual_shape_2d->get_parent() : visual_shape_2d, "remove_child", light_occluder_2d_instance);
```
Simpler.

KoBeWi commented on editor/plugins/visual_shape_2d_editor_plugin.cpp:
```diff
@@ -0,0 +1,292 @@
+		visual_shape_editor->options->show();
+	} else {
+		visual_shape_editor->options->hide();
+		visual_shape_editor->edit(nullptr);
```
This should be happening automatically already.

KoBeWi reviewed (approved):
The implementation looks good.

kitbdev commented on editor/plugins/visual_shape_2d_editor_plugin.h:
```diff
@@ -0,0 +1,91 @@
+		MENU_OPTION_CREATE_LIGHT_OCCLUDER_2D
+	};
+
+	Menu selected_menu_item;
```
Actually, I can remove this, it's unused.

AThousandShips reviewed (approved):
LGTM

akien-mga commented:
I brought this PR to the attention of more maintainers, which sparked some discussion on how we might want to go one step further and fully implement 2D CSG. @smix8 already had a go at it: #99911.

Could you help review that and assess if that fulfills the same use case as this PR in a better way?

Calinou commented:
> Could you help review that and assess if that fulfills the same use case as this PR in a better way?

I feel 2D CSG is best kept separate, as even if you use a single CSG2D node with no children, it'll still have more overhead than a VisualShape2D node.

Calinou reviewed (commented):
Tested locally, it works as expected.

Some feedback:

- The **Circle** mode doesn't extend the polygon all the way down with odd side counts (this happens regardless of the node's aspect ratio):

https://github.com/user-attachments/assets/9b52c070-735f-42d5-a1dc-4dbbcc8c601c

This also happens with the **Capsule** mode with certain side counts such as `5`, albeit in a less extreme manner.

- The **Capsule** mode has identical looks with some side counts (such as `5` and `6`). I'd expect the number of sides specified to be present on each cap of the shape, instead of being the sum of both top and bottom caps. For reference, this is how it's specified in CapsuleMesh in 3D.

https://github.com/user-attachments/assets/7ac98df2-9a88-4908-a31f-56e83d88e7df

- The **Outline Width** property disables filled shape drawing whenever its value is greater than `0.0`. This means you can't draw a filled shape with an outline with a single node, which is inconsistent with how many other nodes work (such as Label or Label3D).
  - Instead, I suggest adding separate **Filled** (bool), **Color** and **Outline Color** properties, and have the **Outline Width** property control outline thickness drawing without disabling filled shape drawing.

- **Outline Width** could allow a value of `-1.0` to draw pixel-perfect lines that *don't* scale with viewport zoom, although antialiasing wouldn't be supported this way. This behavior would essentially be opposite of `CanvasItem.draw_line()` where `-1.0` draws a 1-pixel thick line that scales with viewport zoom. (For a higher-level API, I feel `1.0` should scale with viewport zoom instead, as it's more likely to match users' expectations.)

- The antialiasing feather doesn't scale with viewport zoom, and there's no way to override its size (which can be used to create blurred shapes). The CanvasItem drawing functions don't expose a way to do this, so we'll have to leave it as-is for now. This could be worth exploring in its own proposal.

- There could be an extra mode to draw a rectangle with rounded corners (similar to StyleBoxFlat but with a Node2D base), but it can be done later.

kitbdev commented:
Thanks for the feedback!
> The Circle mode doesn't extend the polygon all the way down with odd side counts (this happens regardless of the node's aspect ratio):

These are regular polygons, so the points are an equal distance from the center, so I can't really stretch them to reach the bottom. They won't always fit into a square shape due to the math. 
Maybe I can reduce the visual rect height for it? But then it wouldn't match the size property and it would be more complex to set.

> The Capsule mode has identical looks with some side counts (such as 5 and 6). I'd expect the number of sides specified to be present on each cap of the shape, instead of being the sum of both top and bottom caps. For reference, this is how it's specified in CapsuleMesh in 3D.

The problem with this is that the capsule's resolution and the circle's resolution are the same, so I wanted it to act similarly. If it controlled the capsule's cap size, then the default value of 64 would have twice as many vertices as the circle does.
Maybe I should halve it when switching to / from the capsule shape like I do with the width?

>  Instead, I suggest adding separate Filled (bool), Color and Outline Color properties, and have the Outline Width property control outline thickness drawing without disabling filled shape drawing.

I wanted to reduce the amount of properties, but I can do this. 
Should the outline be in front or behind the shape? If it is in front, it matches how it looks now with 2 nodes, and the antialiasing would affect the inner edge. If it is in the back antialiasing would not affect the inner edge, but it would better match Label.

> Outline Width could allow a value of -1.0 to draw pixel-perfect lines that don't scale with viewport zoom, although antialiasing wouldn't be supported this way. This behavior would essentially be opposite of CanvasItem.draw_line() where -1.0 draws a 1-pixel thick line that scales with viewport zoom. (For a higher-level API, I feel 1.0 should scale with viewport zoom instead, as it's more likely to match users' expectations.)

I can add this.
I'm not sure about inverting the behavior from the draw_line methods. If it is negative at all it draws a 1px line that doesn't scale. If I flipped it, then to have a scaling width it would need to be negative, and changing the positive values wouldn't make a difference. Maybe it could be a bool in the inspector instead `unscaled outline` or something and it would hide the outline width property.

> There could be an extra mode to draw a rectangle with rounded corners (similar to StyleBoxFlat but with a Node2D base), but it can be done later.

This would require more parameters for each corner radius to match the StyleBoxFlat, and since this isn't resource-based, it would have to be on the base and then hidden on all other shapes (like the resolution).

Calinou commented:
> Should the outline be in front or behind the shape? If it is in front, it matches how it looks now with 2 nodes, and the antialiasing would affect the inner edge. If it is in the back antialiasing would not affect the inner edge, but it would better match Label.

I'd put the outline in front of the shape.
